impl/spanner/mutations/table_test.go

```diff
@@ -0,0 +1,241 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mutations
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/keytransparency/core/adminserver"
+	"github.com/google/keytransparency/core/integration/storagetest"
+	"github.com/google/keytransparency/core/keyserver"
+	"github.com/google/keytransparency/core/water"
+	"github.com/google/keytransparency/impl/spanner/directory"
+	"github.com/google/keytransparency/impl/spanner/testutil"
+	"github.com/google/trillian/crypto/keyspb"
+	"google.golang.org/protobuf/proto"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	dtype "github.com/google/keytransparency/core/directory"
+	ktspanner "github.com/google/keytransparency/impl/spanner"
+	tpb "github.com/google/trillian"
+)
+
+func mustMarshal(t *testing.T, p proto.Message) []byte {
+	t.Helper()
+	b, err := proto.Marshal(p)
+	if err != nil {
+		t.Errorf("proto.Marshal(): %v", err)
+	}
+	return b
+}
+
+func NewForTest(ctx context.Context, t testing.TB, dirID string, logIDs ...int64) *Table {
+	t.Helper()
+	ddl, err := ktspanner.ReadDDL()
+	if err != nil {
+		t.Fatal(err)
+	}
+	client := testutil.CreateDatabase(ctx, t, ddl)
+	q := New(client)
+
+	if err := directory.New(client).Write(ctx, &dtype.Directory{
+		DirectoryID: dirID,
+		Map:         &tpb.Tree{},
+		Log:         &tpb.Tree{},
+		VRFPriv:     &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+	}); err != nil {
+		t.Fatalf("directories.Write(%v): %v", dirID, err)
+	}
+	if err := q.AddLogs(ctx, dirID, logIDs...); err != nil {
+		t.Fatalf("AddLogs(): %v", err)
+	}
+	time.Sleep(readStaleness)
+
+	return q
+}
+
+func TestNewForTest(t *testing.T) {
+	ctx := context.Background()
+	NewForTest(ctx, t, "NewForTest", 1)
+}
+
+func TestMutationLogsIntegration(t *testing.T) {
+	storagetest.RunMutationLogsTests(t,
+		func(ctx context.Context, t *testing.T, dirID string, logIDs ...int64) keyserver.MutationLogs {
+			return NewForTest(ctx, t, dirID, logIDs...)
+		})
+}
+
+func TestLogsAdminIntegration(t *testing.T) {
+	storagetest.RunLogsAdminTests(t,
+		func(ctx context.Context, t *testing.T, dirID string, logIDs ...int64) adminserver.LogsAdmin {
+			return NewForTest(ctx, t, dirID, logIDs...)
+		})
+}
+
+func TestReadBatch(t *testing.T) {
+	const dirID = "readbatch"
+	ctx := context.Background()
+	logID := int64(5)
+	m := NewForTest(ctx, t, dirID, logID)
+	var lastTS water.Mark
+	var err error
+	for i := byte(0); i < 10; i++ {
+		entry := &pb.EntryUpdate{Mutation: &pb.SignedEntry{Entry: mustMarshal(t, &pb.Entry{Index: []byte{i}})}}
+		lastTS, err = m.Send(ctx, dirID, logID, entry, entry, entry)
```

Comment:
I spent over a minute trying to work out why reading one thing led to 3 results. The fact that this has entry repeated is obvious when you see it, and subtle when you don't. Can we pop a comment here that simply says "Write the same entry multiple times to construct a batch" or similar?

Reply:
Added a "batch" variable to clarify this.  I'll also add a todo to remove the variable size param